crates/interface-types/src/lib.rs

```diff
@@ -300,42 +344,110 @@ fn abi2ast(param: &ir::AbiParam) -> Result<ast::WebidlScalarType> {
+    unsafe fn get_memory(&mut self) -> Result<&mut [u8]>;
+}
 
-    let mut copy = |alloc_func_name: &str, bytes: &[u8]| {
-        let len = i32::try_from(bytes.len()).map_err(|_| format_err!("length overflow"))?;
+struct RawTranslateContext<'a> {
+    cx: &'a mut Context,
+    handle: &'a mut InstanceHandle,
+}
+
+impl<'a> RawTranslateContext<'a> {
+    fn new(cx: &'a mut Context, handle: &'a mut InstanceHandle) -> Self {
+        RawTranslateContext { cx, handle }
+    }
+}
+
+impl TranslateContext for RawTranslateContext<'_> {
+    fn invoke_alloc(&mut self, alloc_func_name: &str, len: i32) -> Result<i32> {
         let alloc_args = vec![RuntimeValue::I32(len)];
-        let results = match cx.invoke(handle, alloc_func_name, &alloc_args)? {
+        let results = match self.cx.invoke(self.handle, alloc_func_name, &alloc_args)? {
             ActionOutcome::Returned { values } => values,
             ActionOutcome::Trapped { message } => bail!("trapped: {}", message),
         };
         if results.len() != 1 {
             bail!("allocator function wrong number of results");
         }
-        let ptr = match results[0] {
+        Ok(match results[0] {
             RuntimeValue::I32(i) => i,
             _ => bail!("allocator function bad return type"),
-        };
-        let memory = handle
+        })
+    }
+    unsafe fn get_memory(&mut self) -> Result<&mut [u8]> {
+        let memory = self
+            .handle
             .lookup("memory")
             .ok_or_else(|| format_err!("no exported `memory`"))?;
         let definition = match memory {
             wasmtime_runtime::Export::Memory { definition, .. } => definition,
             _ => bail!("export `memory` wasn't a `Memory`"),
         };
+        Ok(slice::from_raw_parts_mut(
+            (*definition).base,
+            (*definition).current_length,
+        ))
+    }
+}
+
+struct InstanceTranslateContext(pub api::HostRef<api::Instance>);
+
+impl TranslateContext for InstanceTranslateContext {
+    fn invoke_alloc(&mut self, alloc_func_name: &str, len: i32) -> Result<i32> {
+        let alloc = self
+            .0
+            .borrow()
+            .find_export_by_name(alloc_func_name)
+            .ok_or_else(|| format_err!("failed to find alloc function `{}`", alloc_func_name))?
+            .func()
+            .ok_or_else(|| format_err!("`{}` is not a (alloc) function", alloc_func_name))?
+            .clone();
+        let alloc_args = vec![api::Val::I32(len)];
+        let results = match alloc.borrow().call(&alloc_args) {
+            Ok(values) => values,
+            Err(trap) => bail!("trapped: {:?}", trap),
+        };
+        if results.len() != 1 {
+            bail!("allocator function wrong number of results");
+        }
+        Ok(match results[0] {
+            api::Val::I32(i) => i,
+            _ => bail!("allocator function bad return type"),
+        })
+    }
+    unsafe fn get_memory(&mut self) -> Result<&mut [u8]> {
```

Comment:
Would it be possible to make this function safe or encapsulate the unsafety perhaps?

Reply:
`unsafe` is a "guard", so a user will be cautious about memory being a growing one

Reply:
Would it be possible to encapsulate that though? Something perhaps that returns a `Vec<u8>` if that's what happens at the call site anyway?